drivers/clock_control/nrf_power_clock.c

```diff
@@ -286,197 +339,33 @@ static inline void power_event_cb(nrf_power_event_t event)
  */
 void nrf_power_clock_isr(void *arg)
 {
-	u8_t pof, hf_intenset, hf, lf_intenset, lf;
-#if NRF_CLOCK_HAS_CALIBRATION
-	u8_t ctto, done;
-	struct device *dev = arg;
-#endif
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
-	bool usb_detected, usb_pwr_rdy, usb_removed;
-#endif
-
-	pof = (NRF_POWER->EVENTS_POFWARN != 0);
+	ARG_UNUSED(arg);
 
-	hf_intenset = ((NRF_CLOCK->INTENSET &
-		       CLOCK_INTENSET_HFCLKSTARTED_Msk) != 0);
-	hf = (NRF_CLOCK->EVENTS_HFCLKSTARTED != 0);
+	pof_isr();
 
-	lf_intenset = ((NRF_CLOCK->INTENSET &
-		       CLOCK_INTENSET_LFCLKSTARTED_Msk) != 0);
-	lf = (NRF_CLOCK->EVENTS_LFCLKSTARTED != 0);
+	if (clock_event_check_and_clean(NRF_CLOCK_EVENT_HFCLKSTARTED,
+					NRF_CLOCK_INT_HF_STARTED_MASK)) {
+		struct device *hfclk_dev = DEVICE_GET(clock_nrf5_m16src);
+		struct nrf_clock_control *data = hfclk_dev->driver_data;
 
-#if NRF_CLOCK_HAS_CALIBRATION
-	done = (NRF_CLOCK->EVENTS_DONE != 0);
-	ctto = (NRF_CLOCK->EVENTS_CTTO != 0);
-#endif
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
-	usb_detected = nrf_power_event_check(NRF_POWER_EVENT_USBDETECTED);
-	usb_pwr_rdy = nrf_power_event_check(NRF_POWER_EVENT_USBPWRRDY);
-	usb_removed = nrf_power_event_check(NRF_POWER_EVENT_USBREMOVED);
-#endif
-
-	__ASSERT_NO_MSG(pof || hf || hf_intenset || lf
-#if NRF_CLOCK_HAS_CALIBRATION
-			|| done || ctto
-#endif
-#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
-			|| usb_detected || usb_pwr_rdy || usb_removed
-#endif
-	);
-
-	if (pof) {
-		NRF_POWER->EVENTS_POFWARN = 0;
-	}
-
-	if (hf) {
-		NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
-	}
-
-	if (hf_intenset && (hf || ((NRF_CLOCK->HFCLKSTAT &
-				    (CLOCK_HFCLKSTAT_STATE_Msk |
-				     CLOCK_HFCLKSTAT_SRC_Msk)) ==
-				   (CLOCK_HFCLKSTAT_STATE_Msk |
-				    CLOCK_HFCLKSTAT_SRC_Msk)))){
-		/* INTENSET is used as state flag to start calibration,
-		 * hence clear it here.
-		 */
-		NRF_CLOCK->INTENCLR = CLOCK_INTENCLR_HFCLKSTARTED_Msk;
-
-#if defined(CONFIG_SOC_SERIES_NRF52X)
-		/* NOTE: Errata [192] CLOCK: LFRC frequency offset after
-		 * calibration.
-		 * Calibration start, workaround.
-		 */
-		*(volatile u32_t *)0x40000C34 = 0x00000002;
-#endif /* CONFIG_SOC_SERIES_NRF52X */
-
-#if NRF_CLOCK_HAS_CALIBRATION
-		/* Start Calibration */
-		NRF_CLOCK->TASKS_CAL = 1;
-#endif
-	}
-
-	if (lf) {
-		NRF_CLOCK->EVENTS_LFCLKSTARTED = 0;
-
-		if (lf_intenset) {
-			/* INTENSET is used as state flag to start calibration,
-			 * hence clear it here.
-			 */
-			NRF_CLOCK->INTENCLR = CLOCK_INTENCLR_LFCLKSTARTED_Msk;
-
-#if NRF_CLOCK_HAS_CALIBRATION
-			/* Start HF Clock if LF RC is used. */
-			if ((NRF_CLOCK->LFCLKSRCCOPY & CLOCK_LFCLKSRCCOPY_SRC_Msk) ==
-			    CLOCK_LFCLKSRCCOPY_SRC_RC) {
-				ctto = 1U;
-			}
-#endif
+		if (!data->started) {
```

Comment:
I think I already asked about this, but my question somehow got lost.
Is is done because of the anomaly 201? It should be commented properly then. Now it looks like a redundant check.

Reply:
added comment about anomaly 201.